### PR TITLE
Feature/state router base3.0

### PR DESCRIPTION
**Pull request by panxiaojun233**
## What is the purpose of the change

Added an extension of the state router routing chain to improve routing performance

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java [hunk 1]
```diff
@@ -41,11 +66,35 @@
...
+
+    protected URL url;
+
+    protected AtomicReference<AddrCache> cache = new AtomicReference<>();
+    private Semaphore loopPermit = new Semaphore(1);
+
+    private final static ScheduledExecutorService executorService = new ScheduledThreadPoolExecutor(1,
+        new NamedInternalThreadFactory("dubbo-state-router-scheduled-", true));
+
+    private final static ExecutorService loopThreadPool = new ThreadPoolExecutor(1, 1,
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:80]: `loopThreadPool`和`poolRouterThreadPool` 可以合并成一个吗

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java [hunk 2]
```diff
@@ -111,6 +217,94 @@ private void sort() {
...
+                poolRouterThreadPool.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        RouterCache routerCache = null;
+                        try {
+                            routerCache = poolRouter(stateRouter, origin, copyInvokers, notify);
+                            //file cache
+                            newCache.getCache().put(stateRouter.getName(), routerCache);
+                        } catch (Throwable t) {
+                            logger.error("Failed to pool router: " + stateRouter.getUrl() + ", cause: " + t.getMessage(), t);
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:245]: 这里如果失败了，是否不该使用newCache替换原有cache？

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java [hunk 3]
```diff
@@ -111,6 +217,94 @@ private void sort() {
...
+                });
+            }
+        }
+        try {
+            cdl.await();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+
+        this.cache.set(newCache);
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:259]: 如上

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/state/AbstractStateRouter.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.common.extension.ExtensionLoader;
+import org.apache.dubbo.rpc.Invoker;
+import org.apache.dubbo.rpc.cluster.governance.GovernanceRuleRepository;
+
+public abstract class AbstractStateRouter implements StateRouter {
+    protected int priority = DEFAULT_PRIORITY;
+    protected boolean force = false;
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/state/AbstractStateRouter.java:13]: 是否会有线程安全问题，换成 Atomic?

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/tag/TagDynamicStateRouter.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    private TagRouterRule tagRouterRule;
+    private String application;
+
+    public TagDynamicStateRouter(URL url) {
+        super(url);
+        this.priority = TAG_ROUTER_DEFAULT_PRIORITY;
+    }
+
+    @Override
+    public synchronized void process(ConfigChangedEvent event) {
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/tag/TagDynamicStateRouter.java:67]: 这里应该需要 setForce(true)

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/tag/TagStaticStateRouter.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Override
+    public boolean isEnable() {
+        return true;
+    }
+
+    @Override
+    public boolean isForce() {
+        // FIXME
```
Comment (guohao) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/tag/TagStaticStateRouter.java:88]: Fix

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/BitList.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ */
+public class BitList<E> implements BitListInterf<E> {
+    BitListInterf<E> delegate;
+
+    public BitList(List<E> unmodifiableList, boolean empty) {
+        if (Jvm.isJava8Plus()) {
+            delegate = new RoaringBitList<>(unmodifiableList, empty);
+        } else {
+            delegate = new NormalList<>(unmodifiableList, empty);
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/BitList.java:18]: Dubbo3只适用于 Java8+，不需要NormalList

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/Jvm.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
+package org.apache.dubbo.common.utils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ *
+ */
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/Jvm.java:8]: 这个类可以删除，理由同上

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NormalList.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+
+/**
+ *
+ */
+public class NormalList<E> implements BitListInterf<E> {
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NormalList.java:14]: 这个类可以删除，理由同上

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/RoaringBitList.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import org.roaringbitmap.IntIterator;
+import org.roaringbitmap.RoaringBitmap;
+
+class RoaringBitList<E> implements BitListInterf<E> {
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/RoaringBitList.java:12]: 看一下高版本的RoaringBitMap，是否有内置函数支持更高效的遍历

**Review by guohao — CHANGES_REQUESTED** [review]

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java (hunk 1 above)
Comment (panxiaojun233) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/RouterChain.java:80]: 之前好像提过，可能会死锁？

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/BitList.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ *
+ */
+public class BitList<E> implements BitListInterf<E> {
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/BitList.java:11]: 就一个实现类，不需要BitListInterf这个接口拉

**Review by guohao — CHANGES_REQUESTED** [review]

**Review by guohao — COMMENTED** [review]
@chickenlj Pls review this PR

**Pull request closed without merging**